### PR TITLE
[query] refactor math in linear_regression_rows

### PR DESCRIPTION
**Pull request by patrick-schultz**
This makes the implementation match more closely the way I derive the math.

### CODE REVIEW

**Comment by johnc1231** [comment]
Thanks, definitely easier to read. Will thoroughly check tomorrow. Think it's worth benchmarking?

File: hail/python/test/hail/methods/test_statgen.py [hunk 1]
```diff
@@ -476,24 +476,24 @@ def test_weighted_linear_regression(self):
...
-        mt = mt.annotate_cols(y=hl.coalesce(pheno[mt.s].Pheno, 1.0))
+        mt = mt.annotate_cols(y=hl.coalesce(pheno[mt.s].Pheno, 1.0), weight=(mt.col_idx % 2) + 1, weight2=3 * (mt.col_idx % 3) + 1)
         mt = mt.annotate_entries(x=hl.coalesce(mt.GT.n_alt_alleles(), 1.0))
         my_covs = [1.0] + list(covariates[mt.s].values())
 
         ht_with_weights = hl._linear_regression_rows_nd(y=mt.y,
                                           x=mt.x,
                                           covariates=my_covs,
-                                          weights=mt.col_idx)
+                                          weights=mt.weight)
```
Comment (patrick-schultz) [inline hail/python/test/hail/methods/test_statgen.py:486]: I was getting test failures with the old weights. I think what was happening is that using col_idx as a weight created a very badly conditioned regression problem, so different math that's equivalent in exact arithmetic could lead to substantially different results in floating point.

**Comment by patrick-schultz** [comment]
Yeah, I guess I should benchmark, since you've observed some strange brittleness in the linreg performance before.

**Comment by johnc1231** [comment]
I think that brittleness is gone now, and was fixed by us doing method
splitting in compiler. But double checking would be great.

**Review by johnc1231 — DISMISSED** [review]
This is a great change. I'd just like you to run benchmarks to verify that the performance doesn't degrade. If you want, I could run the benchmarks instead.

**Comment by patrick-schultz** [comment]
Just ran benchmarks locally, and it looks fine. Think there's any need to run on the cloud as well?

I also tried benchmarking without the binds, and that didn't appear to make any difference, so I'm going to get rid of those.

File: hail/python/hail/methods/statgen.py [hunk 2]
```diff
@@ -516,20 +513,20 @@ def log_message(i):
...
             kept_samples=ht.kept_samples,
             __scaled_y_nds=ht.scaled_y_nds,
             __sqrt_weight_nds=ht.sqrt_weights,
             ns=ht.ns,
             ds=ht.ns.map(lambda n: n - k - 1),
-            __cov_Qts=ht.cov_Qts,
-            __Qtys=ht.Qtys,
-            __yyps=hl.range(num_y_lists).map(lambda i: dot_rows_with_themselves(ht.scaled_y_nds[i].T) - dot_rows_with_themselves(ht.Qtys[i].T)))
+            __cov_Qs=ht.cov_Qs,
+            __yyps=hl.range(num_y_lists).map(
```
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/python/hail/methods/statgen.py:528]: No I don't, and I agree.

**Comment by johnc1231** [comment]
Locally is sufficient, and yeah get rid of the binds. The binds were totally a case of us luckily rearranging generated code in a way that the JIT liked when we weren't method splitting.

**Comment by johnc1231** [comment]
Also, you're failing linting

**Comment by patrick-schultz** [comment]
I'm closing for now. After more careful benchmarking, this is a slight regression. I think this is the right approach, but we don't have enough facilities for writing performant linear algebra in python/hail yet.

**Pull request closed without merging**